lib/resty/lmdb/transaction.lua

```diff
@@ -0,0 +1,299 @@
+local _M = {}
+
+
+local ffi = require("ffi")
+local base = require("resty.core.base")
+local table_new = require("table.new")
+local table_clear = require("table.clear")
+
+
+ffi.cdef([[
+typedef unsigned int 	MDB_dbi;
+
+
+typedef enum {
+    NGX_LMDB_OP_GET = 0,
+    NGX_LMDB_OP_SET,
+    NGX_LMDB_OP_DB_OPEN,
+    NGX_LMDB_OP_DB_DROP,
+} ngx_lua_resty_lmdb_operation_e;
+
+
+typedef struct {
+    ngx_lua_resty_lmdb_operation_e  opcode;
+    ngx_str_t                       key;   /* GET, SET */
+    ngx_str_t                       value; /* GET, SET */
+    MDB_dbi                         dbi;   /* ALL OPS */
+    unsigned int                    flags; /* SET, DROP */
+} ngx_lua_resty_lmdb_operation_t;
+
+
+int ngx_lua_resty_lmdb_ffi_execute(ngx_lua_resty_lmdb_operation_t *ops,
+    size_t n, int need_write, unsigned char *buf, size_t buf_len, char **err);
+]])
+
+
+local err_ptr = base.get_errmsg_ptr()
+local get_string_buf = base.get_string_buf
+local get_string_buf_size = base.get_string_buf_size
+local setmetatable = setmetatable
+local assert = assert
+local math_max = math.max
+local type = type
+local C = ffi.C
+local ffi_string = ffi.string
+local ffi_new = ffi.new
+local MIN_OPS_N = 16
+local DEFAULT_VALUE_BUF_SIZE = 16 * 1024 -- 16KB
+base.set_string_buf_size(DEFAULT_VALUE_BUF_SIZE)
+local NGX_ERROR = ngx.ERROR
+local NGX_AGAIN = ngx.AGAIN
+local NGX_OK = ngx.OK
+local _TXN_MT = {}
+_TXN_MT.__index = _TXN_MT
+
+
+local CACHED_DBI = {}
+local MDB_CREATE = 0x40000
+local DEFAULT_DB = "_default"
+
+
+local get_dbi
+do
+    function _M.begin(hint)
+        hint = hint or 4
+
+        local txn = table_new(hint, 5)
+
+        txn.n = 0
+        txn.capacity = 0
+        txn.write = false
+        txn.ops_capacity = 0
+        txn.ops = nil
+
+        return setmetatable(txn, _TXN_MT)
+    end
+
+
+    local CACHED_TXN_DBI = _M.begin(1)
+    function _M.get_dbi(create, db)
+        local dbi = CACHED_DBI[db]
+        if dbi then
+            print("cached: ", dbi)
+            return dbi
+        end
+
+        CACHED_TXN_DBI:reset()
+        CACHED_TXN_DBI:db_open(create, db)
+        local res, err = CACHED_TXN_DBI:commit()
+        if not res then
+            return nil, err
+        end
+
+        dbi = CACHED_TXN_DBI[1].result
+
+        -- dbi already cached by commit()
+
+        return dbi
+    end
+    get_dbi = _M.get_dbi
+end
+
+
+function _TXN_MT:reset()
+    self.n = 0
+    self.write = false
+end
+
+
+function _TXN_MT:_new_op()
+    local n = self.n + 1
+    self.n = n
+
+    if n < self.capacity then
+        return self[n]
+    end
+
+    local op = table_new(0, 4)
+    self[n] = op
+    self.capacity = self.n
```

Comment:
Is it necessary to keep a `.capacity` field?

```suggestion
    local op = self[n]
    if not op then
        op = table_new(0, 4)
        self[n] = op
    end
```